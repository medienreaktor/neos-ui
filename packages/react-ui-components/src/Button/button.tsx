import mergeClassNames from 'classnames';
import React from 'react';

import {Omit, PickDefaultProps} from '../../types';
import {makeFocusNode} from '../_lib/focusNode';

export type ButtonStyle = 'clean' | 'brand' | 'lighter' | 'transparent' | 'warn';
export type ButtonHoverStyle = 'clean' | 'brand' | 'darken' | 'warn';
export type ButtonSize = 'small' | 'regular';

interface ButtonTheme {
    readonly 'btn': string;
    readonly 'btn--clean': string;
    readonly 'btn--lighter': string;
    readonly 'btn--transparent': string;
    readonly 'btn--brand': string;
    readonly 'btn--brandActive': string;
    readonly 'btn--brandHover': string;
    readonly 'btn--cleanHover': string;
    readonly 'btn--isPressed': string;
    readonly 'btn--darkenHover': string;
}

// We omit the standard HTML button style attribute,
// so we have no collision with the Button component's style property,
// while still enjoying the intellisense and type checking for the rest of the HTML button attributes
export type HTMLButtonElementAttributesExceptStyle = Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'style'>;

// own props and (optional) HTML button attributes except 'style'
export interface ButtonProps extends HTMLButtonElementAttributesExceptStyle {
    /**
     * This prop controls the visual pressed state of the `Button`.
     */
    readonly isPressed?: boolean;

    /**
     * This prop controls the visual focused state of the `Button`.
     * When `true`, the node gets focused via the DOM API.
     */
    readonly isFocused?: boolean;

    /**
     * This prop controls the visual and interactive disabled state of the `Button`.
     * When `true`, the node gets rendered with a truthy `disabled` prop.
     */
    readonly isDisabled?: boolean;

    /**
     * This prop controls the visual active state of the `Button`.
     */
    readonly isActive?: boolean;

    /**
     * The `kind` prop defines the regular visual style of the `Button`.
     */
    readonly style?: ButtonStyle;

    /**
     * As the `style` prop, this prop controls the visual :hover style of the `Button`.
     */
    readonly hoverStyle?: ButtonHoverStyle;

    /**
     * Defines the size of the button.
     */
    readonly size: ButtonSize;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     * The contents to be rendered within the `Bar`.
     */
    readonly children: React.ReactNode;

    /**
     * An optional css theme to be injected.
     */
    readonly theme?: ButtonTheme;

    /**
     * An interal prop for testing purposes, do not set this prop manually.
     */
    readonly _refHandler?: (isFocused: boolean) => (node: any) => void;
}

type DefaultProps = PickDefaultProps<ButtonProps,
    '_refHandler' |
    'hoverStyle' |
    'isActive' |
    'isDisabled' |
    'isFocused' |
    'size' |
    'style' |
    'type'
>;

export const defaultProps: DefaultProps = {
    _refHandler: makeFocusNode,
    hoverStyle: 'brand',
    isActive: false,
    isDisabled: false,
    isFocused: false,
    size: 'regular',
    style: 'lighter',
    type: 'button',
};

class Button extends React.PureComponent<ButtonProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element {
        const {
            children,
            className,
            isPressed,
            isFocused,
            isDisabled,
            isActive,
            style,
            hoverStyle,
            size,
            theme,
            type,
            _refHandler,
            ...rest
        } = this.props;
        const effectiveStyle = isActive ? 'brand' : style;
        const effectiveHoverStyle = isActive ? 'brand' : hoverStyle;
        const finalClassName = mergeClassNames(
            theme!.btn,
            // @ts-ignore implizit any because ButtonTheme has no index signature
            theme![`btn--size-${size}`],
            // @ts-ignore implizit any because ButtonTheme has no index signature
            theme![`btn--${effectiveStyle!}`],
            // @ts-ignore implizit any because ButtonTheme has no index signature
            theme![`btn--${effectiveHoverStyle!}Hover`],
            {
                [theme!['btn--brandActive']]: isActive,
                [theme!['btn--isPressed']]: isPressed,
            },
            className,
        );

        return (
            <button {...rest} disabled={isDisabled} type={type} className={finalClassName} role="button" ref={_refHandler && _refHandler(isFocused!)}>
                {children}
            </button>
        );
    }
}

export default Button;
