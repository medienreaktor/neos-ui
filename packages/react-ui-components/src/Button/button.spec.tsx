import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import Button, {ButtonProps, defaultProps} from './button';

describe('<Button/>', () => {
    const props: ButtonProps = {
        ...defaultProps,
        children: 'Foo children',
        className: 'foo className',
        hoverStyle: 'clean',
        onClick: () => null,
        style: 'clean',
        theme: {
            'btn': 'btnClassName',
            'btn--brand': 'brandClassName',
            'btn--brandActive': 'brandActiveClassName',
            'btn--brandHover': 'brandHoverClassName',
            'btn--clean': 'cleanClassName',
            'btn--cleanHover': 'cleanHoverClassName',
            'btn--darkenHover': 'darkenHoverClassName',
            'btn--isPressed': 'isPressedClassName',
            'btn--lighter': 'lighterClassName',
            'btn--transparent': 'transparentClassName',
        },
    };

    it('should render correctly.', () => {
        const wrapper = shallow(<Button {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should allow the propagation of "className" with the "className" prop.', () => {
        const wrapper = shallow(<Button {...props} className="fooClassName"/>);

        expect(wrapper.prop('className')).toContain('fooClassName');
    });

    it('should allow the propagation of "type" with the "type" prop.', () => {
        const wrapper = shallow(<Button {...props} type="submit"/>);

        expect(wrapper.prop('type')).toContain('submit');
    });

    it('should allow the propagation of additional props to the wrapper.', () => {
        const wrapper = shallow(<Button {...props} autoFocus={true}/>);

        expect(wrapper.prop('autoFocus')).toBe(true);
    });

    it('should always render the "brand" and "brandHover" theme classNames in case the "isActive" prop is truthy.', () => {
        const wrapper = shallow(<Button {...props} isActive={true}/>);

        expect(wrapper.hasClass(props.theme!['btn--clean'])).toBeFalsy();
        expect(wrapper.hasClass(props.theme!['btn--cleanHover'])).toBeFalsy();
        expect(wrapper.hasClass(props.theme!['btn--brand'])).toBeTruthy();
        expect(wrapper.hasClass(props.theme!['btn--brandHover'])).toBeTruthy();
    });

    it('should render the "style" and "hoverStyle" theme classNames in case the "isActive" prop is falsy.', () => {
        const wrapper = shallow(<Button {...props} isActive={false}/>);

        expect(wrapper.hasClass(props.theme!['btn--clean'])).toBeTruthy();
        expect(wrapper.hasClass(props.theme!['btn--cleanHover'])).toBeTruthy();
        expect(wrapper.hasClass(props.theme!['btn--brand'])).toBeFalsy();
        expect(wrapper.hasClass(props.theme!['btn--brandHover'])).toBeFalsy();
    });

    it('should not render the disabled attribute when passing a falsy "isDisabled" prop.', () => {
        const wrapper = shallow(<Button {...props} isActive={false}/>);

        expect(wrapper.html()).not.toContain('disabled=""');
    });

    it('should render the disabled attribute when passing a truthy "isDisabled" prop.', () => {
        const wrapper = shallow(<Button {...props} isActive={true}/>);

        expect(wrapper.html()).not.toContain('disabled=""');
    });

    it('should call the "_refHandler" prop with the current "isFocused" prop when rendering the node.', () => {
        const _refHandler = jest.fn();

        shallow(<Button {...props} isFocused={false} _refHandler={_refHandler}/>);

        expect(_refHandler.mock.calls.length).toBe(1);
        expect(_refHandler.mock.calls[0][0]).toBe(false);
    });
});
