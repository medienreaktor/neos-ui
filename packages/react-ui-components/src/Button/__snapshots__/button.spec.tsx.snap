// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should render correctly. 1`] = `
<button
  className="btnClassName cleanClassName cleanHoverClassName foo className"
  disabled={false}
  onClick={[Function]}
  role="button"
  type="button"
>
  Foo children
</button>
`;
